content/en/docs/contributing/coding-conventions.md

```diff
@@ -0,0 +1,815 @@
+
+* Check how Doxygen renders macro comments and then decide on the best approach to format.
+
+### Definitions
+
+#### Variable Definitions
+
+Variables are defined at the beginning of the block.
+The block is separated from the code with a blank line.
+
+```c
+int my_func(void)
+{
+    int *ptr;
+    int i, num;
+
+    ptr = malloc(10 * sizeof(int));
+    if (unlikely(!ptr))
+        return -ENOMEM;
+
+    for (i = 0; i < 10; i++)
+        num = /* ... */;
+}
+✅
+```
+
+```c
+int my_func(void)
+{
+    void *ptr = malloc(10 * sizeof(int));
+    
+    if (unlikely(!ptr))
+        return -ENOMEM;
+
+    for (int i = 0; i < 10; i++)
+        int num = /* ... */;
+}
+❌
+```
+
+#### Naming
+
+Identifiers in Unikraft use `snake case`:
+
+```c
+void my_func(void); /* Snake Case  */ ✅
+```
+
+```c
+void myFunc(void);  /* Camel Case  */ ❌
+```
+
+```c
+void MyFunc(void);  /* Pascal Case */ ❌
+```
+
+An exception to this are constants and preprocessor macros, which should use all uppercase and words separated by underscores:
+
+```c
+const int MAGIC_VALUE = 0xdeadb0b0;
+```
+
+```c
+#define IS_DEAD(bobo) ((bobo) == MAGIC_VALUE)
+```
+
+Macro names may also be regular function names if they are just wrappers for actual functions (see `uk_vprintk()` below) or if the functionality might be either implemented as macro or function (e.g., depending on architecture).
+
+### Leading Underscores
+
+A leading single underscore is **discouraged** but may be used selectively to mark identifiers which should not be used directly, like in the header `<uk/print.h>`:
+
+```c
+#if CONFIG_LIBUKDEBUG_PRINTK
+/* please use the uk_printk(), uk_vprintk() macros because they compile in the
+ * function calls only if the configured debug level requires it
+ */
+void _uk_vprintk(int lvl, const char *libname, const char *srcname,
+                 unsigned int srcline, const char *fmt, va_list ap);
+
+
+#define uk_vprintk(lvl, fmt, ap) \
+    do { \
+        if ((lvl) <= KLVL_MAX)  \
+            _uk_vprintk((lvl), __STR_LIBNAME__, __STR_BASENAME__, \
+            __LINE__, (fmt), ap); \
+    } while (0)
+```
+
+Prefer using `UK_PAGE_SIZE` instead of `_PAGE_SIZE` if required to avoid a symbol clash.
+
+However, leading underscores should **not** be used to name private functions within the same source file.
+The `static` keyword already provides the same information.
+
+```c
+static void _i_am_private(int arg)
+{
+        /* ... */
+}
+
+void i_am_public(int arg)
+{
+        _i_am_private(arg);
+}
+❌
+```
+
+Leading double underscores may only be used for:
+
+* low-level native Unikraft types like `__u64` and `__spinlock`
+* header inclusion guards
+* preprocessor macros to add another level of indirection for macro resolution
+
+{{% alert theme="info" %}}
+Rule of thumb: include `libc` headers first and then Unikraft headers.
+{{% /alert %}}
+
+* Use `uk_nonnull` instead of and `UK_NULL` instead of `__NULL`.
```

Comment:
What are `uk_nonnull`, `UK_NULL` and so on? Can't find any of this in our codebase.

Reply:
We don't want to use underscore-starting symbols, due to potential clashes with libc/compiler symbols. So, instead of `__NULL` we use `UK_NULL`. And instead of `__nonnull` we use `uk_nonnull`.

@RaduNichita, you missed a "`__nonnull`" in the above description.